tracer/src/Datadog.Trace/ClrProfiler/AutoInstrumentation/Kafka/KafkaHelper.cs

```diff
@@ -232,46 +232,28 @@ private static long GetMessageSize<T>(T message)
-                                           : $"direction:in,group:{groupId},topic:{topic},type:kafka";
-                        message.Headers.Add(DataStreamsPropagationHeaders.TemporaryEdgeTags, Encoding.UTF8.GetBytes(edgeTags));
-                    }
-                    else
+                    // TODO: we could pool these arrays to reduce allocations
+                    // NOTE: the tags must be sorted in alphabetical order
+                    var edgeTags = string.IsNullOrEmpty(topic)
+                                       ? new[] { "direction:in", $"group:{groupId}", "type:kafka" }
+                                       : new[] { "direction:in", $"group:{groupId}", $"topic:{topic}", "type:kafka" };
+
+                    span.SetDataStreamsCheckpoint(
+                        dataStreamsManager,
+                        CheckpointKind.Consume,
+                        edgeTags,
+                        message is null ? 0 : GetMessageSize(message),
+                        tags.MessageQueueTimeMs == null ? 0 : (long)tags.MessageQueueTimeMs,
+                        pathwayContext);
+
+                    message?.Headers?.Remove(DataStreamsPropagationHeaders.TemporaryBase64PathwayContext); // remove eventual junk
+                    if (!tracer.Settings.KafkaCreateConsumerScopeEnabledInternal && message?.Headers is not null && span.Context.PathwayContext != null)
                     {
-                        // TODO: we could pool these arrays to reduce allocations
-                        // NOTE: the tags must be sorted in alphabetical order
-                        var edgeTags = string.IsNullOrEmpty(topic)
-                                           ? new[] { "direction:in", $"group:{groupId}", "type:kafka" }
-                                           : new[] { "direction:in", $"group:{groupId}", $"topic:{topic}", "type:kafka" };
-
-                        span.SetDataStreamsCheckpoint(
-                            dataStreamsManager,
-                            CheckpointKind.Consume,
-                            edgeTags,
-                            message is null ? 0 : GetMessageSize(message),
-                            tags.MessageQueueTimeMs == null ? 0 : (long)tags.MessageQueueTimeMs,
-                            pathwayContext);
+                        // write the _new_ pathway (the "consume" checkpoint that we just set above) to the headers as a way to pass its value to an eventual
+                        // call to SpanContextExtractor.Extract by a user who'd like to re-pair pathways after a batch consume.
+                        // Note that this header only exists on the consume side, and Kafka never sees it.
+                        var base64PathwayContext = Convert.ToBase64String(BitConverter.GetBytes(span.Context.PathwayContext.Value.Hash.Value));
+                        message.Headers.Add(DataStreamsPropagationHeaders.TemporaryBase64PathwayContext, Encoding.UTF8.GetBytes(base64PathwayContext));
```

Comment:
if, for some weird reason, the header already has that header it will through. It would be safer to a remove first, like we had before?
```suggestion
                        message.Headers.Remove(DataStreamsPropagationHeaders.TemporaryBase64PathwayContext);
                        message.Headers.Add(DataStreamsPropagationHeaders.TemporaryBase64PathwayContext, Encoding.UTF8.GetBytes(base64PathwayContext));
```

Reply:
yes, I'm doing it just above the `if` (I want to remove it even if I cannot extract a new one, because I think it never makes sense for it to be present before this method).

Reply:
Oops, missed that, sorry!